### PR TITLE
GUI widgets: float/int input values can now be modified using a combination of key shortcuts

### PR DESCRIPTION
**Pull request by Sabrina85**

### CODE REVIEW

File: src/odemis/gui/comp/text.py [hunk 1]
```diff
@@ -932,7 +974,9 @@ def __init__(self, *args, **kwargs):
         kwargs['validator'] = IntegerValidator(min_val, max_val, choices, unit)
 
         if 'key_step' not in kwargs and (min_val != max_val):
```
Comment (pieleric) [inline src/odemis/gui/comp/text.py]: If the user passed `key_step_min`, you don't want to override it, so it should be:
`if 'key_step' not in kwargs and 'key_step_min' not in kwargs and (min_val != max_val):`

File: src/odemis/gui/comp/text.py [hunk 2]
```diff
@@ -903,7 +944,8 @@ def __init__(self, *args, **kwargs):
         choices = kwargs.pop('choices', None)
 
         kwargs['validator'] = IntegerValidator(min_val, max_val, choices)
-        kwargs['key_step'] = kwargs.get('key_step', 1)
+        kwargs['key_step'] = kwargs.get('key_step', 1)  # TODO also replace with fallback?
```
Comment (pieleric) [inline src/odemis/gui/comp/text.py]: Yes, I think it should also use `key_step_min` by default, mostly for the sake of all the classes to behave the same by default.

File: src/odemis/gui/comp/text.py [hunk 3]
```diff
@@ -986,7 +1030,8 @@ def __init__(self, *args, **kwargs):
         choices = kwargs.pop('choices', None)
 
         kwargs['validator'] = FloatValidator(min_val, max_val, choices)
-        kwargs['key_step'] = kwargs.get('key_step', 0.1)
+        kwargs['key_step'] = kwargs.get('key_step', 0.1)  # TODO also replace with fallback?
```
Comment (pieleric) [inline src/odemis/gui/comp/text.py]: Yes, that one too.

File: src/odemis/gui/conf/util.py [hunk 4]
```diff
@@ -768,6 +768,9 @@ def text_get(ctrl=value_ctrl, va=va):
             'choices': choices,
         }
 
+        if 'key_step' in conf:
+            ctrl_conf['key_step'] = conf['key_step']
```
Comment (pieleric) [inline src/odemis/gui/conf/util.py]: Let's do the same with `key_step_min`, so that it's fully customisable.

**Comment by Sabrina85** [comment]
I found a bug: It will not bind the method if we provide a "key_step" input explicitly. I fixed it in a new commit. Will do the rest in a separate one.

File: src/odemis/gui/comp/text.py [hunk 5]
```diff
@@ -643,7 +646,7 @@ def __init__(self, *args, **kwargs):
         if val is not None:
             self.SetValue(self._number_value)
 
-        if self.key_step is not None:
+        if self.key_step_fallback is not None:
```
[the remark this answers is not in the document]
Comment (Sabrina85) [reply src/odemis/gui/comp/text.py]: ok you also found my bug :)

File: src/odemis/gui/comp/text.py (hunk 1 above)
Comment (Sabrina85) [reply src/odemis/gui/comp/text.py]: Mhm the user should not pass key_step_min I thought. It should only be calculated if key_step was not provided. That is why I made this new key...

File: src/odemis/gui/comp/text.py (hunk 2 above)
Comment (Sabrina85) [reply src/odemis/gui/comp/text.py]: But should it be the calculated version based on max/min or just replace the string as it is?

File: src/odemis/gui/comp/text.py (hunk 1 above)
Comment (Sabrina85) [reply src/odemis/gui/comp/text.py]: ignore my comment

File: src/odemis/gui/comp/text.py (hunk 2 above)
Comment (pieleric) [reply src/odemis/gui/comp/text.py]: I think you can do exactly the same as the other class, which displays units.

**Comment by Sabrina85** [comment]
Merged the new update with the old one (see second commit) - should contain all requests for changes

File: src/odemis/gui/comp/text.py [hunk 6]
```diff
@@ -778,6 +810,33 @@ def on_char(self, evt):
...
+            # In practice, this results in computing the magnitude of a "little bit" smaller value.
+            value *= (1 + k / 10)
+        try:
+            magnitude = int(math.floor(math.log10(abs(value))))
+            if magnitude <= -15:
+                raise ValueError("Value is so small, so set it zero.")
+        except ValueError:
+            return math.copysign(self.key_step_min, k)
+        step = (10 ** magnitude) * k
+        if self._num_type is int and abs(step) < 1:  # if integer, change at least by 1
```
Comment (pieleric) [inline src/odemis/gui/comp/text.py]: Could we just write something like this?
```
   if self.key_step_min and abs(step) < self.key_step_min:
        step = math.copysign(self.key_step_min, step)
```
Then, num_type can be dropped, and we just pass `key_step_min=1` by default on the *IntText classes?

Comment (Sabrina85) [reply src/odemis/gui/comp/text.py]: I think that code would then overwrite all steps that are smaller then key_step_min though it is possible to do smaller steps. Key_step_min is calculated based on the min/max values, so for the trigger delay it was always 1e-5. However, we would like to allow steps smaller than that. Or do I miss something?
I agree the key_step_min + num_type is not really nice yet...

**Pull request merged**